web/src/components/users/utils.js

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) [2024] SUSE LLC
+ *
+ * All Rights Reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of version 2 of the GNU General Public License as published
+ * by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, contact SUSE LLC.
+ *
+ * To contact SUSE LLC about this file by physical or electronic mail, you may
+ * find current contact information at www.suse.com.
+ */
+
+/**
+ * Method which generates username suggestions based on first and last name.
+ * The method cleans the input name by removing non-alphanumeric characters (except spaces),
+ * splits the name into parts, and then generates suggestions based on these parts.
+ *
+ * @param {string} fullName The full name used to generate username suggestions.
+ * @returns {string[]} An array of username suggestions.
+ */
+const suggestUsernames = (fullName) => {
+  // Cleaning the name.
+  const cleanedName = fullName
+    .normalize('NFD')
+    .trim()
+    .replace(/[\u0300-\u036f]/g, '') // Replacing accented characters with English equivalents, eg. š with s.
+    .replace(/[^\p{L}\p{N} ]/gu, "") // Keep only letters, numbers and spaces. Covering the whole Unicode range, not just ASCII.
+    .toLowerCase();
+
+  // Split the cleaned name into parts.
+  const nameParts = cleanedName.split(/\s+/);
+
+  const suggestions = [];
+
+  nameParts.forEach((namePart, index) => {
+    if (index === 0) {
+      suggestions.push(namePart);
+      suggestions.push(namePart[0]);
+      suggestions.push(namePart[0]);
+      suggestions.push(namePart);
+      suggestions.push(namePart[0]);
+      suggestions.push(namePart);
+    } else {
+      if (index === 1)
+        suggestions[1] += namePart;
+      suggestions[2] += namePart;
+      suggestions[3] += namePart[0];
+      if (index === nameParts.length - 1)
+        suggestions[4] += namePart;
+      else
+        suggestions[4] += namePart[0];
+      suggestions[5] += namePart;
+    }
+  });
```

Comment:
Most probably it's me, but I found this algorithm difficult to follow (even though it can be the most efficient one, I don't know)

What do you think about going for something more verbose like 

```js
//const name = "Maria del Carmen Fernandez Vega";
const suggestions = new Set();
const parts = cleanedName.split(/\s+/);
const firstLetters = parts.map(p => p[0]).join('');
const lastPosition = parts.length - 1;

const [firstPart, ...allExceptFirst] = parts;
const [firstLetter, ...allExceptFirstLetter] = firstLetters;
const lastPart = parts[lastPosition];

// Just the first part of the name
suggestions.add(firstPart);
// The first letter of the first part plus all other parts
suggestions.add(firstLetter + allExceptFirst.join(''));
// The first part plus the first letters of all other parts
suggestions.add(firstPart + allExceptFirstLetter.join(''));
// The first letters except the last one plus the last part
suggestions.add(firstLetters.substring(0, lastPosition) + lastPart);
// All parts without spaces
suggestions.add(parts.join(''));

// let's drop suggestions with less than 3 characters
suggestions.forEach(s => {
  if(s.length < 3) suggestions.delete(s)
});

return [...suggestions];
```

Just an idea. Any other suggestion or improvement about either, the one you proposed and this rough example are also welcome. Specially about naming :P

Reply:
> (even though it can be the most efficient one, I don't know)

I was right and, although your approach contains more entries (e.g. _mdel_) than mine, it is more efficient in terms of performance, see https://jsben.ch/VuEsA.

So, we need another view here :)


@imobachgs, @jilopez. Friendly ping :)

Reply:
If I have to choose, then I would vote for the second one. The difference in performance is insignificant and the code is much easier to read, understand and maintain. Moreover, this method will be executed only a couple of times. IMHO nothing to worry here about performance.